File: _audit_i18n.js
// Точный аудит переводов: реально вычисляет объект I18N и сравнивает ключи.
const fs = require('fs');
const path = require('path');
const src = fs.readFileSync(path.join(__dirname, 'index.html'), 'utf8');

const m = src.match(/const I18N = (\{[\s\S]*?\n\});/);
const I18N = eval('(' + m[1] + ')');

const langs = Object.keys(I18N);
const all = new Set();
langs.forEach(l => Object.keys(I18N[l]).forEach(k => all.add(k)));
console.log('Всего уникальных ключей:', all.size);

let problems = 0;
langs.forEach(l => {
  const miss = [...all].filter(k => !(k in I18N[l]));
  console.log(`[${l}] ключей: ${Object.keys(I18N[l]).length}  пропущено: ${miss.length ? miss.join(', ') : '— нет —'}`);
  if (miss.length) problems++;
});

// HTML data-i18n покрытие
const htmlKeys = [...src.matchAll(/data-i18n(?:-html)?="([^"]+)"/g)].map(x => x[1]);
const trset = new Set(Object.keys(I18N.tr));
const badHtml = [...new Set(htmlKeys)].filter(k => !trset.has(k));
console.log(`\n[HTML data-i18n] уникальных: ${new Set(htmlKeys).size}  нет в переводах: ${badHtml.length ? badHtml.join(', ') : '— нет —'}`);
if (badHtml.length) problems++;

// t('literal') с границей слова (чтобы не путать с openSheet/get)
const tk = [...src.matchAll(/(?<![\w$])t\('([a-zA-Z_]\w*)'\)/g)].map(x => x[1]);
const badT = [...new Set(tk)].filter(k => !trset.has(k));
console.log(`[JS t('...')] литеральных: ${new Set(tk).size}  нет в переводах: ${badT.length ? badT.join(', ') : '— нет —'}`);
if (badT.length) problems++;

console.log(problems ? `\n⚠️ Проблем: ${problems}` : '\n✅ Все переводы на месте, во всех трёх языках.');

File: _audit_i18n.py
"""Аудит переводов: сравнивает ключи tr/ru/en в I18N, ищет пропуски."""
import re, os

path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "index.html")
src = open(path, encoding="utf-8").read()

# вырезаем блок const I18N = { ... };
m = re.search(r"const I18N = \{(.*?)\n\};", src, re.S)
block = m.group(1)

# делим на языковые секции: tr: { ... }, ru: { ... }, en: { ... }
langs = {}
for lang in ["tr", "ru", "en"]:
    lm = re.search(r"\n  %s: \{(.*?)\n  \}" % lang, block, re.S)
    body = lm.group(1)
    # ключи = идентификатор в начале строки (с отступом) перед двоеточием
    keys = re.findall(r"\n    ([a-zA-Z_]\w*)\s*:", body)
    langs[lang] = keys

all_keys = set()
for l in langs:
    all_keys |= set(langs[l])

print("Всего уникальных ключей:", len(all_keys))
for lang in ["tr", "ru", "en"]:
    s = set(langs[lang])
    missing = sorted(all_keys - s)
    dups = sorted(k for k in set(langs[lang]) if langs[lang].count(k) > 1)
    print("\n[%s] ключей: %d" % (lang, len(s)))
    print("  ПРОПУЩЕНО:", missing if missing else "— нет —")
    if dups:
        print("  ДУБЛИ КЛЮЧЕЙ:", dups)

# проверка: все data-i18n / data-i18n-html в HTML существуют в tr
html_keys = set(re.findall(r'data-i18n(?:-html)?="([^"]+)"', src))
tr_keys = set(langs["tr"])
bad_html = sorted(k for k in html_keys if k not in tr_keys)
print("\n[HTML data-i18n] использовано ключей:", len(html_keys))
print("  НЕТ В ПЕРЕВОДАХ:", bad_html if bad_html else "— нет —")

# статические t('xxx') в JS (без конкатенации) — извлекаем литералы
t_keys = set(re.findall(r"t\('([a-zA-Z_]\w*)'\)", src))
bad_t = sorted(k for k in t_keys if k not in tr_keys)
print("\n[JS t('...')] литеральных ключей:", len(t_keys))
print("  НЕТ В ПЕРЕВОДАХ:", bad_t if bad_t else "— нет —")
